File: Two Pointers/977. Squares of a Sorted Array/977. Squares of a Sorted Array.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var sortedSquares = function (nums) {
  let ans = [];
  let l = 0;
  let r = nums.length - 1;

  while (l <= r) {
    const low_squared = nums[l] ** 2;
    const high_squared = nums[r] ** 2;
    if (low_squared > high_squared) {
      ans.push(low_squared);
      l++;
    } else {
      ans.push(high_squared);
      r--;
    }
  }

  return ans.reverse();
};

// Time: O(n)
// Space: O(n)

File: Two Pointers/977. Squares of a Sorted Array/977. Squares of a Sorted Array.py
class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        ans = []
        l = 0
        r = len(nums) - 1

        while l <= r:
            low_squared = nums[l] ** 2
            high_squared = nums[r] ** 2
            if low_squared > high_squared:
                ans.append(low_squared)
                l += 1
            else:
                ans.append(high_squared)
                r -= 1

        ans.reverse()
        return ans

# Time: O(n)
# Space: O(n)